File: app/flutter/lib/api/group.dart
import 'dart:convert';

import 'package:logging/logging.dart';

import 'package:betro/encryption//betro_dart_lib.dart';

import './auth.dart';
import './types/GroupResponse.dart';
import './types/CreateGroupRequest.dart';

final _logger = Logger('api/follow');

class GroupController {
  final AuthController auth;
  GroupController(this.auth);

  Future<List<GroupResponse>> fetchGroups() async {
    final response = await auth.client.get('/api/groups');
    return (response.data as List<dynamic>)
        .map((a) => GroupResponse.fromJson(a))
        .toList();
  }

  Future<bool> deleteGroup(String groupId) async {
    final response = await auth.client.delete('/api/groups/$groupId');
    return response.data['deleted'];
  }

  Future<GroupResponse?> createGroup(
    String name, [
    bool is_default = false,
  ]) async {
    final encryptionKey = auth.encryptionKey;
    if (encryptionKey == null) return null;
    final symKey = generateSymKey();
    final encryptedSymKey = await symEncrypt(
      encryptionKey,
      base64Decode(symKey),
    );
    final req = CreateGroupRequest(
      sym_key: encryptedSymKey,
      name: name,
      is_default: is_default,
    );
    final response = await auth.client.post(
      '/api/groups',
      data: jsonEncode(req),
    );
    _logger.finer(response.data);
    return GroupResponse.fromJson(response.data);
  }
}

File: libraries/client/src/group.ts
import { Buffer } from "buffer";
import { symEncrypt, generateSymKey } from "@repo/encryption/src";

import AuthController from "./auth";
import { GroupResponse } from "./types";

class GroupController {
  auth: AuthController;
  constructor(auth: AuthController) {
    this.auth = auth;
  }
  fetchGroups = async (): Promise<Array<GroupResponse> | null> => {
    try {
      const response = await this.auth.instance.get("/api/groups");
      const data = response.data;
      return data;
    } catch (e) {
      return null;
    }
  };

  deleteGroup = async (
    groupId: string
  ): Promise<{
    deleted: boolean;
  } | null> => {
    try {
      const response = await this.auth.instance.delete(
        `/api/groups/${groupId}`
      );
      const data = response.data;
      return data;
    } catch (e) {
      return null;
    }
  };

  createGroup = async (
    name: string,
    is_default: boolean
  ): Promise<GroupResponse | null> => {
    const sym_key = await generateSymKey();
    const encryptedSymKey = await symEncrypt(
      this.auth.encryptionKey,
      Buffer.from(sym_key, "base64")
    );
    try {
      const response = await this.auth.instance.post("/api/groups", {
        name: name,
        sym_key: encryptedSymKey,
        is_default: is_default,
      });
      return response.data;
    } catch (e) {
      return null;
    }
  };
}

export default GroupController;
